cpp/src/arrow/adapters/avro/adapter.h

```diff
@@ -0,0 +1,93 @@
+// regarding copyright ownership.  The ASF licenses this file
+// to you under the Apache License, Version 2.0 (the
+// "License"); you may not use this file except in compliance
+// with the License.  You may obtain a copy of the License at
+//
+//   http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+#ifndef ARROW_AVRO_CONVERTER_H
+#define ARROW_AVRO_CONVERTER_H
+
+#include <cstdint>
+#include <memory>
+#include <string>
+
+#include "arrow/builder.h"
+#include "arrow/buffer.h"
+#include "arrow/memory_pool.h"
+#include "arrow/status.h"
+#include "arrow/table.h"
+#include "arrow/type.h"
+#include "arrow/type_traits.h"
+#include "arrow/util/bit-util.h"
+#include "arrow/util/macros.h"
+#include "arrow/util/visibility.h"
+
+#include "avro.h"
+
+namespace arrow {
+
+namespace adapters {
+
+namespace avro {
+
+/// Base class for reading from an avro datasource.
+//
+/// This class provides a facilities for incrementally building the null bitmap
+/// (see Append methods) and as a side effect the current number of slots and
+/// the null count.
+class ARROW_EXPORT AvroArrowReader {
+ public:
+  explicit AvroArrowReader(MemoryPool* pool)
+      : pool_(pool) {
+  }
+
+  virtual ~AvroArrowReader() = default;
+
+  Status ReadFromAvroFile(avro_file_reader_t* file_reader,
+                          std::shared_ptr<Array>* out);
+ protected:
+  AvroArrowReader() {
+  }
+  ;
+  MemoryPool* pool_;
+
+  Status GenerateAvroSchema(avro_file_reader_t* file_reader,
+                            std::shared_ptr<DataType>* out);
+
+  std::shared_ptr<DataType> get_arrow_type(avro_schema_t schema);
+
+  Status generic_read(const avro_value_t val,
+                      std::shared_ptr<ArrayBuilder> builder);
+
+  Status read_string(const avro_value_t val, std::shared_ptr<ArrayBuilder> builder);
+  Status read_binary(const avro_value_t val, std::shared_ptr<ArrayBuilder> builder);
+  Status read_fixed(const avro_value_t val, std::shared_ptr<ArrayBuilder> builder);
+  Status read_int32(const avro_value_t val, std::shared_ptr<ArrayBuilder> builder);
+  Status read_int64(const avro_value_t val, std::shared_ptr<ArrayBuilder> builder);
+  Status read_float64(const avro_value_t val, std::shared_ptr<ArrayBuilder> builder);
+  Status read_float32(avro_value_t val, std::shared_ptr<ArrayBuilder> builder) ;
+  Status read_bool(const avro_value_t val, std::shared_ptr<ArrayBuilder> builder);
+  Status read_array(const avro_value_t val, std::shared_ptr<ArrayBuilder> builder);
+  Status read_map(const avro_value_t val, std::shared_ptr<ArrayBuilder> builder);
+  Status read_record(const avro_value_t val, std::shared_ptr<ArrayBuilder> builder);
```

Comment:
You'll want all of these particular implementations to be available for inlining in the private implementation. Additionally, you should pass the builder as `ArrayBuilder*` as passing by value will call the copy ctor

Reply:
If you look in https://github.com/apache/arrow/pull/802 it allocates builders for the schema and then provides them as bare pointers so we can avoid the overhead of interacting with `std::shared_ptr` for each value that we append